src/vos/vos_dtx.c

```diff
@@ -717,6 +730,520 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
 	return rc;
 }
 
+struct vos_dtx_share {
+	/** Link into the daos_tx_handle::dth_shares */
+	d_list_t		vds_link;
+	/** The DTX record type, see enum vos_dtx_record_types. */
+	uint32_t		vds_type;
+	/** The record in the related tree in SCM. */
+	umem_id_t		vds_record;
+};
+
+static bool
+vos_dtx_is_normal_entry(struct umem_instance *umm, umem_id_t entry)
+{
+	if (UMMID_IS_NULL(entry) ||
+	    umem_id_equal(umm, entry, DTX_UMMID_ABORTED) ||
+	    umem_id_equal(umm, entry, DTX_UMMID_UNKNOWN))
+		return false;
+
+	return true;
+}
+
+static int
+vos_dtx_alloc(struct umem_instance *umm, struct daos_tx_handle *dth,
+	      umem_id_t rec_mmid, struct vos_dtx_entry_df **dtxp)
+{
+	daos_key_t		*dkey = dth->dth_dkey;
+	struct vos_dtx_entry_df	*dtx;
+	struct vos_container	*cont;
+	umem_id_t		 dtx_mmid;
+	daos_iov_t		 kiov;
+	daos_iov_t		 riov;
+	struct dtx_rec_bundle	 rbund;
+	int			 rc;
+
+	cont = vos_hdl2cont(dth->dth_coh);
+	D_ASSERT(cont != NULL);
+
+	dtx_mmid = umem_zalloc(umm, sizeof(struct vos_dtx_entry_df));
+	if (UMMID_IS_NULL(dtx_mmid))
+		return -DER_NOMEM;
+
+	dtx = umem_id2ptr(umm, dtx_mmid);
+	dtx->te_xid = dth->dth_xid;
+	dtx->te_oid = dth->dth_oid;
+	if (dkey != NULL) {
+		dtx->te_dkey_hash[0] = d_hash_murmur64(dkey->iov_buf,
+						       dkey->iov_len,
+						       VOS_BTR_MUR_SEED);
+		dtx->te_dkey_hash[1] = d_hash_string_u32(dkey->iov_buf,
+							 dkey->iov_len);
+	} else {
+		dtx->te_dkey_hash[0] = 0;
+		dtx->te_dkey_hash[1] = 0;
+	}
+	dtx->te_epoch = dth->dth_epoch;
+	dtx->te_ver = dth->dth_ver;
+	dtx->te_state = DTX_ST_INIT;
+	dtx->te_flags = 0;
+	dtx->te_intent = dth->dth_intent;
+	dtx->te_sec = time(NULL);
+	dtx->te_records = rec_mmid;
+	dtx->te_next = UMMID_NULL;
+	dtx->te_prev = UMMID_NULL;
+
+	rbund.trb_ummid = dtx_mmid;
+	daos_iov_set(&riov, &rbund, sizeof(rbund));
+	daos_iov_set(&kiov, &dth->dth_xid, sizeof(dth->dth_xid));
+	rc = dbtree_upsert(cont->vc_dtx_active_hdl, BTR_PROBE_EQ,
+			   DAOS_INTENT_UPDATE, &kiov, &riov);
+	if (rc == 0) {
+		dth->dth_ent = dtx_mmid;
+		*dtxp = dtx;
+	}
```

Comment:
else free?

Reply:
The DTX will be auto freed when the PMDK transaction is aborted.